instance.stop - don't deactivate purged child objects

alena1108 opened the pull request:
And LB fixes

    1) fixed the tests to always wait for LB instance to come to final state,
     otherwise simulator cleanup can result in error
    2) lb.removeHost - don't trigger LB-agent instance cleanup if the host
    itself is in removing/removed state. Instance cleanup will be handled by
    generic host cleanup process where all instances on host are marked as
    removed
    3) LB config update - create LB-agent instance(s) only on active hosts

alena1108 commented:
@ibuildthecloud Darren, can you pls review. This PR takes care of Lb integration test fixes (see more info in the first comment)

ibuildthecloud commented:
Let me think this through a bit more.  The logic seems brittle when we start looking for specific states.

ibuildthecloud commented:
What exactly does this fix?  Do certain errors go away or does it run faster??

alena1108 commented:
@ibuildthecloud there are 3 fixes:

1) When host is removed from the LB as a result of host.remove, don't try to cleanup LB system instance , but rather rely on the default host cleanup logic to mark the instance as removed.I also removed the test checking that lb instance was gone after host removal, because I believe it should be tested in host lifecycle specific test where we verify that all instances on removed host are marked as removed. That test used to take a lot of time because I was waiting for the instance to be gone.

2) When apply haproxy configuration on lb instances, skip the hosts that are in Removed/Removing states

3) The checks I've put to instance.stop handler. This handler is invoked directly form instance.removed, so its quite possible to run into situation when instance failed to remove for the first time (some of its resources were removed/purged); and for a subsequent removal instance.stop will be called, and it will error out when we try to deactivate already purged resources.

ibuildthecloud commented:
LGTM, should I merge?

alena1108 commented:
Sure.
